src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+                    existing.ProtocolVersion = peer.ProtocolVersion ?? existing.ProtocolVersion;
+                    existing.PublicKey = peer.PublicKey ?? existing.PublicKey;
+                    existing.Latency = peer.Latency ?? existing.Latency;
+                    existing.Addresses = existing
+                       .Addresses
+                       .Union(peer.Addresses)
+                       .ToList();
+
+                    return existing;
+                });
+
+            if (!isNew)
+            {
+                return p;
+            }
+
+            if (_log.IsDebugEnabled)
+            {
+                _log.Debug($"New peer registerd {p}");
+            }
+            
+            PeerDiscovered?.Invoke(this, p);
+
+            return p;
+        }
+
+        /// <summary>
+        ///   Deregister a peer.
+        /// </summary>
+        /// <param name="peer">
+        ///   The peer to remove..
+        /// </param>
+        /// <remarks>
+        ///   Remove all knowledge of the peer. The <see cref="PeerRemoved"/> event
+        ///   is raised.
+        /// </remarks>
+        public void DeregisterPeer(Peer peer)
+        {
+            if (peer.Id == null)
+            {
+                throw new ArgumentNullException(nameof(peer));
+            }
+
+            if (_otherPeers.TryRemove(peer.Id.ToBase58(), out var found))
+            {
+                peer = found;
+            }
+            
+            PeerRemoved?.Invoke(this, peer);
+        }
+
+        /// <summary>
+        ///   Determines if a connection is being made to the peer.
+        /// </summary>
+        /// <param name="peer">
+        ///   A <see cref="Peer"/>.
+        /// </param>
+        /// <returns>
+        ///   <b>true</b> is the <paramref name="peer"/> has a pending connection.
+        /// </returns>
+        public bool HasPendingConnection(Peer peer) { return _pendingConnections.TryGetValue(peer, out _); }
+
+        /// <summary>
+        ///   The addresses that cannot be used.
+        /// </summary>
+        public MultiAddressBlackList BlackList { get; set; } = new MultiAddressBlackList();
+
+        /// <summary>
+        ///   The addresses that can be used.
+        /// </summary>
+        public MultiAddressWhiteList WhiteList { get; set; } = new MultiAddressWhiteList();
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="dhtService"></param>
+        public SwarmService(IPeerRouting dhtService = null) { Router = dhtService; }
+
+        /// <inheritdoc />
+        public Task StartAsync()
+        {
+            if (LocalPeer == null)
+            {
+                throw new NotSupportedException("The LocalPeer is not defined.");
+            }
+
+            // Many of the unit tests do not setup the LocalPeerKey.  If
+            // its missing, then just use plaintext connection.
+            // TODO: make the tests setup the security protocols.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)